monomer_test.go

```diff
@@ -1,19 +1,213 @@
+		ParentHash: common.HexToHash("0x2"),
+		GasLimit:   3000000,
+		Hash:       common.HexToHash("0x3"),
+	}
+}
+
+func TestToComet(t *testing.T) {
+	header := newTestHeader()
+	cometHeader := header.ToComet()
+
+	require.Equal(t, &bfttypes.Header{
+		ChainID: header.ChainID.String(),
+		Height:  int64(header.Height),
+		Time:    time.Unix(int64(header.Time), 0),
+		AppHash: header.StateRoot.Bytes(),
+	}, cometHeader)
+}
+
+func TestToEth(t *testing.T) {
+	header := newTestHeader()
+	ethHeader := header.ToEth()
+
+	require.Equal(t, &ethtypes.Header{
+		ParentHash:      header.ParentHash,
+		Root:            header.StateRoot,
+		Number:          big.NewInt(int64(header.Height)),
+		GasLimit:        header.GasLimit,
+		MixDigest:       common.Hash{},
+		Time:            header.Time,
+		UncleHash:       ethtypes.EmptyUncleHash,
+		ReceiptHash:     ethtypes.EmptyReceiptsHash,
+		BaseFee:         common.Big0,
+		WithdrawalsHash: &ethtypes.EmptyWithdrawalsHash,
+		Difficulty:      common.Big0,
+	}, ethHeader)
+}
+
+func TestBlockNewBlock(t *testing.T) {
+	block := monomer.NewBlock(newTestHeader(), bfttypes.Txs{})
+	ethBlock, err := block.ToEth()
+	require.NoError(t, err)
+
+	newBlock, err := monomer.SetHeader(block)
+	require.NoError(t, err)
+	require.Equal(t, ethBlock.Hash(), newBlock.Header.Hash)
+
+	_, err = monomer.SetHeader(nil)
+	require.Error(t, err)
+}
+
+func TestBlockSetHeader(t *testing.T) {
+	block := monomer.NewBlock(newTestHeader(), bfttypes.Txs{})
+	ethBlock, err := block.ToEth()
+	require.NoError(t, err)
+
+	newBlock, err := monomer.SetHeader(block)
+	require.NoError(t, err)
+	require.Equal(t, ethBlock.Hash(), newBlock.Header.Hash)
+
+	_, err = monomer.SetHeader(nil)
+	require.Error(t, err)
+}
+
+func TestBlockMakeBlock(t *testing.T) {
+	header := newTestHeader()
+	emptyTxs := bfttypes.Txs{}
+
+	block := monomer.NewBlock(header, emptyTxs)
+	ethBlock, err := block.ToEth()
+	require.NoError(t, err)
+
+	newBlock, err := monomer.MakeBlock(header, emptyTxs)
+	require.NoError(t, err)
+
+	require.Equal(t, ethBlock.Hash(), newBlock.Header.Hash)
+}
+
 func TestBlockToEth(t *testing.T) {
 	l1InfoTx, depositTx, cosmosEthTx := testutils.GenerateEthTxs(t)
-	block := testutils.GenerateBlockFromEthTxs(t, l1InfoTx, []*ethtypes.Transaction{depositTx}, []*ethtypes.Transaction{cosmosEthTx})
+	block := testutils.GenerateBlockFromEthTxs(t,
+		l1InfoTx,
+		[]*ethtypes.Transaction{depositTx},
+		[]*ethtypes.Transaction{cosmosEthTx},
+	)
+
+	ethTxs, err := rolluptypes.AdaptCosmosTxsToEthTxs(block.Txs)
+	require.NoError(t, err)
+
 	ethBlock, err := block.ToEth()
 	require.NoError(t, err)
-	for i, ethTx := range []*ethtypes.Transaction{l1InfoTx, depositTx, cosmosEthTx} {
-		require.EqualExportedValues(t, ethTx, ethBlock.Body().Transactions[i])
+	require.EqualExportedValues(t, ethtypes.NewBlockWithWithdrawals(
+		block.Header.ToEth(),
+		ethTxs,
+		nil,
+		[]*ethtypes.Receipt{},
+		[]*ethtypes.Withdrawal{},
+		trie.NewStackTrie(nil),
```

Comment:
We're getting pretty close to testing the implementation here. Let's expect the final result, rather than passing to a helper. Something like:

```golang
require.EqualExportedValues(t, &ethtypes.Block{
	Header: block.Header.ToEth(),
	// ... the other fields ...
}, ethBlock)
```

The test for `ToCometLikeBlock` is good. We should mimic that here.

Reply:
ethtypes.Block's fields are private:
```go
type Block struct {
    header       *Header
    uncles       []*Header
    transactions Transactions
    withdrawals  Withdrawals

    // caches
    hash atomic.Value
    size atomic.Value

    // These fields are used by package eth to track
    // inter-peer block relay.
    ReceivedAt   time.Time
    ReceivedFrom interface{}
}
```
When I update 
```suggestion
	require.Equal(t, ethtypes.NewBlockWithWithdrawals(
		block.Header.ToEth(),
		ethTxs,
		nil,
		[]*ethtypes.Receipt{},
		[]*ethtypes.Withdrawal{},
		trie.NewStackTrie(nil),
```
I have an error
```
Error Trace:    /Users/daniilankusin/monomer/monomer_test.go:130
                Error:          Not equal: 
                                expected: &types.Block{header:(*types.Header)(0x140004e2508), uncles:[]*types.Header(nil), transactions:types.Transactions{(*types.Transaction)(0x140003bc070), (*types.Transaction)(0x140003bc150), (*types.Transaction)(0x140003bc1c0)}, withdrawals:types.Withdrawals{}, hash:atomic.Value{v:interface {}(nil)}, size:atomic.Value{v:interface {}(nil)}, ReceivedAt:time.Date(1, time.January, 1, 0, 0, 0, 0, time.UTC), ReceivedFrom:interface {}(nil)}
                                actual  : &types.Block{header:(*types.Header)(0x140004e2008), uncles:[]*types.Header(nil), transactions:types.Transactions{(*types.Transaction)(0x140003bc310), (*types.Transaction)(0x140003bc380), (*types.Transaction)(0x140003bc3f0)}, withdrawals:types.Withdrawals{}, hash:atomic.Value{v:interface {}(nil)}, size:atomic.Value{v:interface {}(nil)}, ReceivedAt:time.Date(1, time.January, 1, 0, 0, 0, 0, time.UTC), ReceivedFrom:interface {}(nil)}
                            
                                Diff:
                                --- Expected
                                +++ Actual
                                @@ -122,4 +122,4 @@
                                    time: (time.Time) {
                                -    wall: (uint64) 13953742264056089080,
                                -    ext: (int64) 50291918,
                                +    wall: (uint64) 13953742264056103080,
                                +    ext: (int64) 50306001,
                                     loc: (*time.Location)({
                                @@ -955,4 +955,4 @@
                                    time: (time.Time) {
                                -    wall: (uint64) 13953742264056096080,
                                -    ext: (int64) 50299209,
                                +    wall: (uint64) 13953742264056104080,
                                +    ext: (int64) 50307084,
                                     loc: (*time.Location)({
                                @@ -1579,4 +1579,4 @@
                                    time: (time.Time) {
                                -    wall: (uint64) 13953742264056098080,
                                -    ext: (int64) 50301126,
                                +    wall: (uint64) 13953742264056105080,
                                +    ext: (int64) 50307751,
                                     loc: (*time.Location)({
                Test:           TestBlockToEth
```
It's impossible to change the time here, so I'm going to rewrite test to handle fields separately.